core/src/main/scala/chisel3/Data.scala

```diff
@@ -893,6 +893,50 @@ abstract class Data extends HasId with NamedComponent with SourceInfoDoc {
   def toPrintable: Printable
 }
 
+object Data {
+  implicit class DataEquality[T <: Data](lhs: T)(implicit sourceInfo: SourceInfo, compileOptions: CompileOptions) {
+    def ===(rhs: T): Bool = {
+      (lhs, rhs) match {
+        case (thiz: UInt, that: UInt) => thiz === that
+        case (thiz: SInt, that: SInt) => thiz === that
+        case (thiz: AsyncReset, that: AsyncReset) => thiz.asBool === that.asBool
+        case (thiz: Reset, that: Reset) => thiz === that
+        case (thiz: Interval, that: Interval) => thiz === that
+        case (thiz: FixedPoint, that: FixedPoint) => thiz === that
+        case (thiz: EnumType, that: EnumType) => thiz === that
+        case (thiz: Clock, that: Clock) => thiz.asUInt === that.asUInt
+        case (thiz: Vec[_], that: Vec[_]) =>
+          if (thiz.length != that.length) {
+            false.B
+          } else {
+            thiz.getElements
+              .zip(that.getElements)
+              .map { case (thisData, thatData) => thisData === thatData }
+              .reduce(_ && _)
+          }
+        case (thiz: Record, that: Record) =>
+          if (thiz.elements.size != that.elements.size) {
+            false.B
+          } else {
+            thiz.elements
+              .zip(that.elements)
+              .map {
+                case ((thisName, thisData), (thatName, thatData)) =>
+                  thisName.equals(thatName).asBool && thisData === thatData
+              }
+              .reduce(_ && _)
+          }
+        // This should be matching to (DontCare, DontCare) but the compiler wasn't happy with that
+        case (_: DontCare.type, _: DontCare.type) => true.B
```

Comment:
I think comparing DontCare should not be possible. We do not want these two versions of code to (silently) have different behavior:

```
DontCare === DontCare
```

```
val a = Wire(UInt(4.W))
val b = Wire(UInt(4.W))
a := DontCare
b := DontCare
a === b

Reply:
> We do not want these two versions of code to (silently) have different behavior

Absolutely agreed, but the behavior of the two actually is the same, both will give `true`. Here's a Scastie illustrating the behavior:
```scala
import chisel3._
import chisel3.util.Valid
import chisel3.experimental.BundleLiterals._

class Example extends Module {

  val out1, out2      = IO(Output(Bool()))

  val u1 = WireInit(UInt(8.W), DontCare)
  val u2 = WireInit(UInt(8.W), DontCare)
  out1 := u1 === u2
  
  val b1 = Valid(UInt(8.W)).Lit(_.bits -> 123.U) // unspecified is implicitly DontCare
  val b2 = Valid(UInt(8.W)).Lit(_.bits -> 123.U)
  out2 := b1.asUInt === b2.asUInt
}
```
This gives the following FIRRTL
```
circuit Example :
  module Example :
    input clock : Clock
    input reset : UInt<1>
    output out1 : UInt<1>
    output out2 : UInt<1>

    wire u1 : UInt<8>
    u1 is invalid
    wire u2 : UInt<8>
    u2 is invalid
    node _out1_T = eq(u1, u2)
    out1 <= _out1_T
    wire _out2_WIRE : UInt<1>
    _out2_WIRE is invalid
    node _out2_T = cat(_out2_WIRE, UInt<7>("h7b"))
    wire _out2_WIRE_1 : UInt<1>
    _out2_WIRE_1 is invalid
    node _out2_T_1 = cat(_out2_WIRE_1, UInt<7>("h7b"))
    node _out2_T_2 = eq(_out2_T, _out2_T_1)
    out2 <= _out2_T_2
```
And the following Verilog:
```verilog
module Example(
  input   clock,
  input   reset,
  output  out1,
  output  out2
);
  assign out1 = 1'h1; // @[main.scala 13:14]
  assign out2 = 1'h1; // @[main.scala 17:21]
endmodule
```

Reply:
I'm not sure this is a good thing, but it is consistent.

Reply:
As annoying as it is, comparing two invalidated wires should return true. A lot of Chisel code intentionally or unintentionally assumes that invalid is zero in certain contexts. This is documented in CIRCT. I would love to change this, but it would be very difficult at this point. I do think defining a new concept of poison and seeing where that goes is tractable.

Reply:
> I do think defining a new concept of poison and seeing where that goes is tractable.

But beyond the scope of this PR for sure 🙂

Reply:
The ambiguity around `DontCare === 0.U` is why I feel like making them _uncomparable_ by generating an elaboration time error might be the best solution here.

Reply:
I agree that the ambiguity is a problem, but is it a better outcome that the following would fail:
```scala
val b1 = Valid(UInt(8.W)).Lit(_.bits -> 123.U) // unspecified is implicitly DontCare
val b2 = Valid(UInt(8.W)).Lit(_.bits -> 123.U)
b1 === b2
```
But this would not?
```scala
val b1 = Valid(UInt(8.W)).Lit(_.bits -> 123.U) // unspecified is implicitly DontCare
val b2 = Valid(UInt(8.W)).Lit(_.bits -> 123.U)
WireInit(b1) === WireInit(b2)
```